torch/distributed/checkpoint/state_dict_saver.py

```diff
@@ -105,21 +107,47 @@ def save(
     """
     torch._C._log_api_usage_once("torch.distributed.checkpoint.save")
 
-    dumpable_state_dict = {}
-    for key, elem in state_dict.items():
-        dumpable_state_dict[key] = (
-            elem.state_dict() if isinstance(elem, Stateful) else elem
-        )
-
     return _save_state_dict(
-        dumpable_state_dict,
+        _stateful(state_dict),
         storage_writer,
         process_group,
         coordinator_rank,
         no_dist,
         planner,
     )
 
+def async_save(state_dict: STATE_DICT_TYPE, *args: Any, **kwargs: Any) -> Future:
+    """Asynchronous version of ``save_state_dict``. This code first de-stages the state_dict on CPU, and then calls
+    `save` in a separate thread.
+
+    .. warning::
+        This feature is experimental and subject to removal/change.
+
+    Returns:
+        Future: A future holding the resultant Metadata object from `save`.
+
+    """
+
+    # since calling nccl collectives in new threads is risky, we call `_stateful` here in case users have
+    # defined custom collectives
+    cpu_state_dict = _offload_state_dict_to_cpu(_stateful(state_dict))
+
+    executor = ThreadPoolExecutor(max_workers=1)
+    f = executor.submit(
+        _save_state_dict,
+        cpu_state_dict,
+        *args,
+        **kwargs
+    )
+    f.add_done_callback(lambda f: executor.shutdown(wait=False))
+
+    return f
+
+def _stateful(state_dict: STATE_DICT_TYPE) -> STATE_DICT_TYPE:
```

Comment:
Should we have a verb in the function name as this converts a Stateful to its state_dict?

Reply:
Personally I think the verb is a little unnecessary. I think a close analogy is a method like `sorted` or `list`, where `get` or `make` wouldn't add a lot of value. Do you have a strong opinion on this?

Reply:
I feel like I kinda stand in between. I don't really have a strong opinion on the name, since it is a private helper util, but in general I would prefer either the name of the helper is self-explanatory or we can have a short docstring describing it.

Reply:
Well, I agree the `list` case. But for the `list` case, the return value is a `list`. Here, the function is called `_stateful` but what it does is actually convert the `Stateful` objects to the state_dict. So we should at least change it to a name with more actual meaning.